alu_u32/src/lt/stark.rs

```diff
@@ -31,48 +31,68 @@ where
             .map(|(bit, base)| bit * base)
             .sum();
 
-        // Check bit decomposition of z = 256 + input_1[n] - input_2[n], where
-        // n is the most significant byte that differs between inputs
-        for i in 0..3 {
-            builder
-                .when_ne(local.byte_flag[i], AB::Expr::one())
-                .assert_eq(local.input_1[i], local.input_2[i]);
+        // check that the n-th byte flag is set, where n is the first byte that differs between the two inputs
 
+        // ensure at most one byte flag is set
+        let flag_sum =
+            local.byte_flag[0] + local.byte_flag[1] + local.byte_flag[2] + local.byte_flag[3];
+        builder.assert_bool(flag_sum.clone());
+        // check that bytes before the first set byte flag are all equal
+        // case: top bytes match
+        builder
+            .when_ne(local.byte_flag[0], AB::Expr::one())
+            .assert_eq(local.input_1[0], local.input_2[0]);
+        // case: top two bytes match
+        builder
+            .when_ne(local.byte_flag[0] + local.byte_flag[1], AB::Expr::one())
+            .assert_eq(local.input_1[1], local.input_2[1]);
+        // case: top three bytes match
+        builder
+            .when_ne(
+                local.byte_flag[0] + local.byte_flag[1] + local.byte_flag[2],
+                AB::Expr::one(),
+            )
+            .assert_eq(local.input_1[2], local.input_2[2]);
+        // case: top four bytes match; must set z = 0
+        builder
+            .when_ne(flag_sum.clone(), AB::Expr::one())
+            .assert_eq(local.input_1[3], local.input_2[3]);
+        builder
+            .when_ne(flag_sum.clone(), AB::Expr::one())
+            .assert_eq(bit_comp.clone(), AB::Expr::zero());
+
+        // Check bit decomposition of z = 256 + input_1[n] - input_2[n]
+        // when `n` is the first byte that differs between the two inputs.
+        for i in 0..4 {
             builder.when(local.byte_flag[i]).assert_eq(
                 AB::Expr::from_canonical_u32(256) + local.input_1[i] - local.input_2[i],
                 bit_comp.clone(),
             );
-
+            // ensure that when the n-th byte flag is set, the n-th bytes are actually different
```

Comment:
This constraint fixes a soundness gap: otherwise, a malicious prover could e.g. set `input_1 = [0, 0, 1, 0]`, `input_2 = [0, 0, 1, 1]`, and set the `byte_flag`'s to `[0, 0, 1, 0]`. Then the other constraints would have the result depend on the comparison of the 3rd byte (which is equal), when in fact the 4th byte determines the correct value of the comparison.